File: _includes/code/howto/manage-data.update.ts
// How-to: Manage data -> Update objects - TypeScript examples
// run with: node --loader=ts-node/esm manage-data.update.ts
import assert from 'assert';

// ================================
// ===== INSTANTIATION-COMMON =====
// ================================
import weaviate, { WeaviateClient } from 'weaviate-client';

const client = await weaviate.connectToWCD(
  'WEAVIATE_INSTANCE_URL',  // Replace WEAVIATE_INSTANCE_URL with your instance URL
 {
   authCredentials: new weaviate.ApiKey('api-key'),
   headers: {
     'X-OpenAI-Api-Key': process.env.OPENAI_API_KEY || '',  // Replace with your inference API key
   }
 } 
)

const collectionName = 'JeopardyQuestion';

// UpdateProps START // Replace START
const myCollection = client.collections.get('JeopardyQuestion') 
// UpdateProps END // Replace END

// ============================
// ===== Define the class =====
// ============================

const collectionDefinition = {
  name: 'JeopardyQuestion',
  description: 'A Jeopardy! question',
  vectorizers: weaviate.configure.vectorizer.text2VecCohere(),
};

// Clean slate
try {
  await client.collections.delete(collectionName)
} catch {
  // Ignore error if class didn't exist
} finally {
  await client.collections.create(collectionDefinition)
}


// =============================
// ===== Update properties =====
// =============================
{
// UpdateProps START
// highlight-start
const response = await myCollection.data.update({
  id: 'ed89d9e7-4c9d-4a6a-8d20-095cb0026f54',
  properties: {
    'points': 100,
  },
})
// highlight-end

console.log(response)
// UpdateProps END
}
// Test
// result = await client.data.getterById().withId(id).withClassName(className).do();
// assert.equal(result.properties['points'], 100);
// assert.equal(result.properties['question'], 'Test question');  // make sure we didn't REPLACE the object


// =========================
// ===== Update vector =====
// =========================
{
// UpdateVector START
const jeopardy = await client.collections.get('Jeopardy')
const response = await jeopardy.data.update({
  id: 'ed89d9e7-4c9d-4a6a-8d20-095cb0026f54',
  // highlight-start
  vectors: Array(1536).fill(0.12345), // new vector value
  // highlight-end
})

console.log(response)
// UpdateVector END
}

// ==========================
// ===== Replace object =====
// ==========================

// Replace START
// highlight-start
const response = await myCollection.data.replace({
  // highlight-end
  id: 'ed89d9e7-4c9d-4a6a-8d20-095cb0026f54',
  // highlight-start
  properties: {
    'answer': 'Replaced',
    // The other properties will be deleted
  },
  // highlight-end
})

console.log(response)
// Replace END

// Test
// result = await client.data.getterById().withId(id).withClassName(className).do();
// assert.deepEqual(result.properties, { answer: 'Replaced' });  // ensure the other props were deleted


// =============================
// ===== Delete properties =====
// =============================

// DelProps START
async function deleteProperties(client: WeaviateClient, uuidToUpdate: string, collectionName: string, propNames: string[]) {
  const collection = client.collections.get(collectionName);
  const objectData = await collection.query.fetchObjectById(uuidToUpdate);
  const propertiesToUpdate = objectData?.properties;
  
  if (propertiesToUpdate) {
    for (let propName of propNames) {
        if (propName in propertiesToUpdate) {
          delete propertiesToUpdate[propName];
        }
    }
  
    result = await collection.data.replace({
      id: uuidToUpdate,
      properties: propertiesToUpdate
    })
  }
}
  
let id = 'ed89d9e7-4c9d-4a6a-8d20-095cb0026f54'
deleteProperties(client, id, 'JeopardyQuestion', ['answer'])
// DelProps END

// Test
// result = await client.data.getterById().withId(id).withClassName(className).do();
// assert.deepEqual(result.properties, {});

File: _includes/code/howto/manage-data.update.py
# How-to: Manage-data -> Update objects - Python examples
import os

# ================================
# ===== INSTANTIATION-COMMON =====
# ================================

import weaviate
import weaviate.classes as wvc
import weaviate_datasets as wd

# Instantiate the client with the OpenAI API key
client = weaviate.connect_to_local(
    headers={
        "X-OpenAI-Api-Key": os.environ["OPENAI_API_KEY"]  # Replace with your inference API key
    }
)

d = wd.WineReviewsNV()
d.upload_dataset(client, overwrite=True)

try:

    # ============================
    # ===== Define the class =====
    # ============================

    # Clean slate
    if client.collections.exists("JeopardyQuestion"):
        client.collections.delete("JeopardyQuestion")

    client.collections.create(
        name="JeopardyQuestion",
        description="A Jeopardy! question",
        vectorizer_config=wvc.config.Configure.Vectorizer.text2vec_openai()
    )

    # =============================
    # ===== Update properties =====
    # =============================

    # UpdateProps START
    uuid = "..."  # replace with the id of the object you want to update

    # UpdateProps END

    # Actually get the ID for testing
    jeopardy = client.collections.get("JeopardyQuestion")

    uuid = jeopardy.data.insert({
        "question": "Test question",
        "answer": "Test answer",
        "points": -1,
    })

    # UpdateProps START
    jeopardy = client.collections.get("JeopardyQuestion")
    jeopardy.data.update(
        uuid=uuid,
        # highlight-start
        properties={
            "points": 100,
        }
        # highlight-end
    )
    # UpdateProps END

    # Test
    result = jeopardy.query.fetch_object_by_id(uuid)
    assert result.properties["points"] == 100

    # =========================
    # ===== Update vector =====
    # =========================
    # UpdateVector START
    jeopardy = client.collections.get("JeopardyQuestion")
    jeopardy.data.update(
        uuid=uuid,
        properties={
            "points": 100,
        },
        # highlight-start
        vector=[0.12345] * 1536
        # highlight-end
    )
    # UpdateVector END

    # Test
    result = jeopardy.query.fetch_object_by_id(uuid, include_vector=True)
    assert len(result.vector["default"]) == 1536
    assert str(result.vector["default"][0])[:5] == str(0.123)


    # ================================
    # ===== Update named vectors =====
    # ================================

    # Commented as updating named vectors is not yet available in the Python client
    # reviews = client.collections.get("WineReviewNV")
    # review_uuid = reviews.query.fetch_objects(limit=1).objects[0].uuid


    # # UpdateNamedVector START
    # reviews = client.collections.get("WineReviewNV")
    # reviews.data.update(
    #     uuid=review_uuid,
    #     properties={
    #         "title": "A delicious wine",
    #         "review_body": "This mystery wine is a delight to the senses.",
    #         "country": "Mordor"
    #     },
    #     # highlight-start
    #     vector={
    #         "title": [0.12345] * 1536,
    #         "review_body": [0.54321] * 1536,
    #         "title_country": [0.050505] * 1536,
    #     }
    #     # highlight-end
    # )
    # # UpdateVector END

    # # Test
    # result = jeopardy.query.fetch_object_by_id(uuid, include_vector=True)
    # for vn in ["title", "review_body", "title_country"]:
    #     assert len(result.vector[vn]) == 1536
    # assert str(result.vector["title_country"][0])[:5] == str(0.050)


    # ==========================
    # ===== Replace object =====
    # ==========================
    # Replace START
    jeopardy = client.collections.get("JeopardyQuestion")
    # highlight-start
    jeopardy.data.replace(
    # highlight-end
        uuid=uuid,
        properties={
            "answer": "Replaced",
            # The other properties will be deleted
        },
    )
    # Replace END

    # Test
    result = jeopardy.query.fetch_object_by_id(uuid)
    assert result.properties["answer"] == "Replaced"
    assert result.properties["points"] == None  # ensure the other props were deleted
    assert result.properties["question"] == None  # ensure the other props were deleted


    # =============================
    # ===== Delete properties =====
    # =============================

    # DelProps START
    from typing import List
    from weaviate import WeaviateClient

    def del_props(client: WeaviateClient, uuid_to_update: str, collection_name: str, prop_names: List[str]) -> None:
        collection = client.collections.get(collection_name)

        # fetch the object to update
        object_data = collection.query.fetch_object_by_id(uuid_to_update)
        properties_to_update = object_data.properties

        # remove unwanted properties
        for prop_name in prop_names:
            if prop_name in properties_to_update:
                del properties_to_update[prop_name]

        # replace the properties
        collection.data.replace(
            uuid=uuid_to_update,
            properties=properties_to_update
        )


    uuid = "..."  # replace with the id of the object you want to delete properties from
    # DelProps END

    uuid = result.uuid  # Get the ID for testing

    # DelProps START
    del_props(client, uuid, "JeopardyQuestion", ["answer"])
    # DelProps END

    # Test
    result = jeopardy.query.fetch_object_by_id(uuid)
    assert result.properties["answer"] == None

finally:
    client.close()
